File: JavaScript/Algorithms/Sorting/bubble_sort.js
const unsortedNumbers = [99, 44, 6, 1, 2, 5, 63, 87, 283, 4, 0]

function bubbleSort(numArr) {
  const length = numArr.length;

  for (let i = 0; i < length; i++) {
    for (let j = 0; j < length; j++) {
      if (numArr[j] > numArr[j + 1]) {
        const temp = numArr[j];
        numArr[j] = numArr[j + 1]
        numArr[j + 1] = temp;
      }
    }
  }

  return numArr;
}

function bubbleSortOptimized(numArr) {
  const length = numArr.length;

  for (let i = 0; i < length; i++) {
    let swapped = false;

    for (let j = 0; j < length; j++) {
      if (numArr[j] > numArr[j + 1]) {
        const temp = numArr[j];
        numArr[j] = numArr[j + 1];
        numArr[j + 1] = temp;
        swapped = true;
      }
    }

    if (!swapped) {
      break;
    }
  }

  return numArr;
}

console.log(bubbleSort(unsortedNumbers));
console.log(bubbleSortOptimized(unsortedNumbers));

File: Python/Algorithms/Sorting Algorithms/bubble_sort.py
numbers = [99, 44, 6, 1, 2, 5, 63, 87, 283, 4, 0]


def bubble_sort(num_arr):
    arr_len = len(num_arr)

    for i in range(arr_len):
        for j in range(0, arr_len - i - 1):
            if num_arr[j] > num_arr[j + 1]:
                temp = num_arr[j]
                num_arr[j] = num_arr[j + 1]
                num_arr[j + 1] = temp

    return num_arr


def bubble_sort_optimized(num_arr):
    arr_len = len(num_arr)

    for i in range(arr_len):
        swapped = False

        for j in range(0, arr_len - i - 1):
            if num_arr[j] > num_arr[j + 1]:
                temp = num_arr[j]
                num_arr[j] = num_arr[j + 1]
                num_arr[j + 1] = temp
                swapped = True

        if not swapped:
            break

    return num_arr


# print(bubble_sort(numbers))
print(bubble_sort_optimized(numbers))
